Common/TableProducer/OccupancyTableProducer.cxx

```diff
@@ -0,0 +1,1159 @@
+///
+/// \brief  TPC PID - Calibration
+/// \author Rahul Verma (rahul.verma@iitb.ac.in) :: Marian I Ivanov (marian.ivanov@cern.ch)
+
+#include "Framework/runDataProcessing.h"
+#include "Framework/AnalysisTask.h"
+#include "Framework/O2DatabasePDGPlugin.h"
+
+#include "Common/DataModel/TrackSelectionTables.h"
+#include "Common/DataModel/PIDResponse.h"
+#include "Common/DataModel/EventSelection.h"
+#include "Common/DataModel/Centrality.h"
+#include "Common/DataModel/Multiplicity.h"
+
+#include "PWGLF/DataModel/LFStrangenessTables.h"
+
+#include "OccupancyTables.h"
+
+#include <typeinfo>
+
+#include "Framework/AnalysisDataModel.h"
+#include "Framework/HistogramRegistry.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "DataFormatsFT0/Digit.h"
+#include "DataFormatsParameters/GRPLHCIFData.h"
+// #include "MetadataHelper.h"
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+int32_t nBCsPerOrbit = o2::constants::lhc::LHCMaxBunches;
+
+struct occTableProducer {
+
+  Service<o2::ccdb::BasicCCDBManager> ccdb;
+
+  // declare production of tables
+  Produces<aod::OccIndexTable> GenOccIndexTable;
+  Produces<aod::Occs> GenOccTable;
+
+  Configurable<int> customOrbitOffset{"customOrbitOffset", 0, "customOrbitOffset for MC"};
+
+  // Histogram registry;
+  //  HistogramRegistry recoTracks  {"recoTracks"  , {}, OutputObjHandlingPolicy::AnalysisObject, true, true};
+  HistogramRegistry recoEvent{"recoEvent", {}, OutputObjHandlingPolicy::AnalysisObject, true, true};
+
+  void init(InitContext const&)
+  {
+    LOGF(info, "Starting init");
+
+    // Getting Info from CCDB, to be implemented
+    //  auto& mgr = o2::ccdb::BasicCCDBManager::instance();
+    //  // http://ccdb-test.cern.ch:8080/browse/Users/r/raverma?report=true
+    //  mgr.setURL("http://ccdb-test.cern.ch:8080");
+    //  mgr.setCaching(true);
+    //  auto myVect = mgr.get<TVectorT<float>>("Users/r/raverma/");
+    //  if( !myVect){
+    //    LOG(info)<<"ERROR ::CCDB OBJECT Not found";
+    //    return;
+    //  }
+
+    recoEvent.add("h_nBCsPerTF", "h_nBCsPerTF", {HistType::kTH1F, {{100, 114040, 114060}}}); // 114048
+    recoEvent.add("h_nBCinTF", "h_nBCinTF", {HistType::kTH1F, {{1000, 0, 1000000}}});
+    recoEvent.add("h_RO_T0V0Prim_Unfm_80", "h_RO_T0V0Prim_Unfm_80:median contributors", {HistType::kTH1F, {{12 * 2, -1, 11}}});
+    recoEvent.add("h_RO_FDDT0V0Prim_Unfm_80", "h_RO_FDDT0V0Prim_Unfm_80:median contributors", {HistType::kTH1F, {{12 * 2, -1, 11}}});
+    recoEvent.add("h_RO_NtrackDet_Unfm_80", "h_RO_NtrackDetITS/TPC/TRD/TOF_80:median contributors", {HistType::kTH1F, {{12 * 2, -1, 11}}});
+
+    LOG(info) << "Printing Event Info ";
+    recoEvent.print();
+    LOG(info) << "Finishing Init ";
+  }
+
+  template <typename T, typename U>
+  int BinarySearchTable(T Key, U Table, int low, int high)
+  {
+    while (low <= high) {
+      int mid = low + (high - low) / 2;
+      if (Key == Table.iteratorAt(mid).trackId()) {
```

Comment:
This creates a temporary table iterator each time, which is not a cheap operation. Instead, create one iterator at the beginning and shift it to the desired position.

Reply:
Is the implemented correct now
     auto midElement = Table.iteratorAt(mid).trackId();
     if (Key == midElement) { changed from => // if (Key == Table.iteratorAt(mid).trackId()) {
        return mid;
      }
      if (Key > midElement) {  changed from => //  if (Key > Table.iteratorAt(mid).trackId()) {
        low = mid + 1;
      } // If Key is greater, ignore left  half, update the low